contracts/Synthswap.sol

```diff
@@ -5,85 +5,295 @@ import "./interfaces/ISynthSwap.sol";
-        synthetix = ISynthetix(_synthetix);
+    constructor (
+        address _sUSD,
+        address _aggregationRouterV4,
+        address _addressResolver,
+        address _volumeRewards,
+        address _treasury
+    ) {
         sUSD = IERC20(_sUSD);
-        volumeRewards = _volumeRewards;
-        aggregationRouterV3 = _aggregationRouterV3;
+        router = IAggregationRouterV4(_aggregationRouterV4);
         addressResolver = IAddressResolver(_addressResolver);
+        volumeRewards = _volumeRewards;
+        treasury = _treasury;
     }
 
+    //////////////////////////////////////
+    ///////// EXTERNAL FUNCTIONS /////////
+    //////////////////////////////////////
+
+    receive() external payable {
+        emit Received(msg.sender, msg.value);
+    }
+
+    /// @inheritdoc ISynthSwap
     function swapInto(
-        bytes calldata _data,
-        bytes32 destinationSynthCurrencyKey
+        bytes32 _destSynthCurrencyKey,
+        address _destSynthAddress,
+        bytes calldata _data
     ) external payable override returns (uint) {
+        (uint amountOut,) = swapOn1inch(_data, false);
 
-        // Make sure to set destReceiver to this contract
-        (bool success, bytes memory returnData) = aggregationRouterV3.delegatecall(_data);
-        require(success, _getRevertMsg(returnData));
-        (uint sUSDAmountOut,) = abi.decode(returnData, (uint, uint));
-        
-        sUSD.approve(address(synthetix), sUSDAmountOut);
-        uint amountReceived = synthetix.exchangeWithTrackingForInitiator(
-            'sUSD', // hardcode source currency key
-            sUSDAmountOut, //source amount
-            destinationSynthCurrencyKey,
-            volumeRewards, 
-            'KWENTA' //tracking code
-        );
-        
-        emit SwapInto(msg.sender, amountReceived);
-        return amountReceived;
+        // if destination synth is NOT sUSD, swap on Synthetix is necessary 
+        if (_destSynthCurrencyKey != sUSD_CURRENCY_KEY) {
+            amountOut = swapOnSynthetix(
+                amountOut, 
+                sUSD_CURRENCY_KEY, 
+                _destSynthCurrencyKey, 
+                address(sUSD)
+            );
+        }
+
+        IERC20(_destSynthAddress).safeTransfer(msg.sender, amountOut);
+  
+        emit SwapInto(msg.sender, amountOut);
+        return amountOut;
     }
 
+    /// @inheritdoc ISynthSwap
     function swapOutOf(
-        bytes32 sourceSynthCurrencyKey,
-        uint sourceAmount, // Make sure synthetix is approved to use this amount
+        bytes32 _sourceSynthCurrencyKey,
+        address _sourceSynthAddress,
+        uint _sourceAmount,
         bytes calldata _data
     ) external override returns (uint) {
+        IERC20(_sourceSynthAddress).safeTransferFrom(msg.sender, address(this), _sourceAmount);
+
+        // if source synth is NOT sUSD, swap on Synthetix is necessary 
+        if (_sourceSynthCurrencyKey != sUSD_CURRENCY_KEY) {
```

Comment:
Technically frontend shouldn't be calling synthswap in this flow. So up to you if you want to keep this `if`

Reply:
Are you saying there will never be a case when a user has sUSD and uses SynthSwap to swap into an ERC20 token/ETH?

Reply:
Correct we would just send that transaction to 1inch directly

Reply:
Okay. Well, the contract should function correctly with or without the front-end, so I think I will leave the check